### PR TITLE
libs/arch_libc:Add kasan protection for memset and other mm api

### PR DESCRIPTION
**Pull request by W-M-R**
## Summary
1. libc/machine:Add prefixes to libc functions implemented by arch
        kasan does not instrument the assembly function,
        so it is checked in advance before calling.
        If Kasan is not turned on, the speed and space will
        be almost unchanged under compiler optimization.
2. libs/machine:strnlen need check return
3. arch_lib.c: close kasan readcheck no return
4. libc: Modify the check length
## Impact
After enabling archlibc, kasan checks will be performed when calling mm api
## Testing
None

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

File: libs/libc/machine/arch_libc.c [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+ ****************************************************************************/
+
+#include <string.h>
+
+/****************************************************************************
+ * Public Functions prototypes
+ ****************************************************************************/
+
+#ifdef CONFIG_LIBC_ARCH_MEMCHR
+FAR void *arch_memchr(FAR const void *s, int c, size_t n);
```
Comment (anchao) [inline libs/libc/machine/arch_libc.c:32]:
```suggestion
FAR void *ARCH_LIBCFUN(memchr)(FAR const void *s, int c, size_t n);
```
replace all arch define to ARCH_LIBCFUN(x) style

**Review by anchao — CHANGES_REQUESTED** [review]

**Review by anchao — APPROVED** [review]
nice.

**Comment by lupyuen** [comment]
Sorry @W-M-R: Could you take a look at the build for `nucleo-f072rb:nsh`, is it failing because of this PR? Thanks!
```text
Configuration/Tool: nucleo-f072rb/nsh,CONFIG_ARM_TOOLCHAIN_CLANG
/tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: sched/libsched.a(sem_destroy.c.obj): in function `nxsem_destroy':
sem_destroy.c:(.text.nxsem_destroy+0x24): undefined reference to `__atomic_compare_exchange_2'
```
https://github.com/NuttX/nuttx/actions/runs/11632461378/job/32395707420#step:7:132

**Comment by W-M-R** [comment]
> Sorry @W-M-R: Could you take a look at the build for `nucleo-f072rb:nsh`, is it failing because of this PR? Thanks!
> 
> ```
> Configuration/Tool: nucleo-f072rb/nsh,CONFIG_ARM_TOOLCHAIN_CLANG
> /tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: sched/libsched.a(sem_destroy.c.obj): in function `nxsem_destroy':
> sem_destroy.c:(.text.nxsem_destroy+0x24): undefined reference to `__atomic_compare_exchange_2'
> ```
> 
> https://github.com/NuttX/nuttx/actions/runs/11632461378/job/32395707420#step:7:132

ok，i  will deal with this patch latter

**Comment by crafcat7** [comment]
> Sorry @W-M-R: Could you take a look at the build for `nucleo-f072rb:nsh`, is it failing because of this PR? Thanks!
> 
> ```
> Configuration/Tool: nucleo-f072rb/nsh,CONFIG_ARM_TOOLCHAIN_CLANG
> /tools/gcc-arm-none-eabi/bin/../lib/gcc/arm-none-eabi/13.2.1/../../../../arm-none-eabi/bin/ld: sched/libsched.a(sem_destroy.c.obj): in function `nxsem_destroy':
> sem_destroy.c:(.text.nxsem_destroy+0x24): undefined reference to `__atomic_compare_exchange_2'
> ```
> 
> https://github.com/NuttX/nuttx/actions/runs/11632461378/job/32395707420#step:7:132

Hi, I have submitted a fix to resolve this issue https://github.com/apache/nuttx/pull/14605.
The problem is that the previously removed macros are brought back, causing arch_atomic.c to fail to compile.

**Comment by lupyuen** [comment]
Cool thank you so much! :-)

**Pull request merged**